File: TCS-CODEVITA/Season-13/Round-1/Stellar-Journey/solution.java
import java.util.*;

public class StellarJourney {
    static class Segment {
        int x1, y1, x2, y2;
        Segment(int a, int b, int c, int d){ x1=a; y1=b; x2=c; y2=d; }
    }

    static class Star {
        int id;
        List<Segment> segments = new ArrayList<>();
        Set<String> points = new HashSet<>();
        Star(int id){ this.id=id; }
        void addSegment(Segment s){
            segments.add(s);
            points.add(s.x1+","+s.y1);
            points.add(s.x2+","+s.y2);
        }
    }

    static boolean sharePoint(Star a, Star b){
        for(String p : a.points){
            if(b.points.contains(p)) return true;
        }
        return false;
    }

    static int bfs(boolean[][] graph, int start, int end){
        Queue<int[]> q = new LinkedList<>();
        boolean[] visited = new boolean[graph.length];
        q.add(new int[]{start,1});
        visited[start]=true;
        while(!q.isEmpty()){
            int[] cur = q.poll();
            int node = cur[0], dist = cur[1];
            if(node==end) return dist;
            for(int i=0;i<graph.length;i++){
                if(graph[node][i] && !visited[i]){
                    visited[i]=true;
                    q.add(new int[]{i,dist+1});
                }
            }
        }
        return -1;
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        Segment[] segs = new Segment[N];
        for(int i=0;i<N;i++){
            int a=sc.nextInt(), b=sc.nextInt(), c=sc.nextInt(), d=sc.nextInt();
            segs[i] = new Segment(a,b,c,d);
        }
        int sx = sc.nextInt(), sy = sc.nextInt();
        int dx = sc.nextInt(), dy = sc.nextInt();

        // Step 1: Group segments into stars (DFS on shared points)
        boolean[] visited = new boolean[N];
        List<Star> stars = new ArrayList<>();
        int starId=0;
        for(int i=0;i<N;i++){
            if(!visited[i]){
                Star s = new Star(starId++);
                dfs(i, segs, visited, s);
                stars.add(s);
            }
        }

        // Step 2: Build connectivity graph
        int nStars = stars.size();
        boolean[][] graph = new boolean[nStars][nStars];
        for(int i=0;i<nStars;i++){
            for(int j=i+1;j<nStars;j++){
                if(sharePoint(stars.get(i), stars.get(j))){
                    graph[i][j]=graph[j][i]=true;
                }
            }
        }

        // Step 3: Identify start and destination stars
        int startStar=-1, endStar=-1;
        for(Star s: stars){
            for(Segment seg: s.segments){
                if(startStar==-1 && pointOnSegment(seg, sx, sy)) startStar = s.id;
                if(endStar==-1 && pointOnSegment(seg, dx, dy)) endStar = s.id;
            }
        }

        if(startStar==-1 || endStar==-1){
            System.out.println("Impossible");
            return;
        }

        // Step 4: BFS to find minimum stars
        int ans = bfs(graph, startStar, endStar);
        if(ans==-1) System.out.println("Impossible");
        else System.out.println(ans);

        sc.close();
    }

    static void dfs(int idx, Segment[] segs, boolean[] visited, Star s){
        if(visited[idx]) return;
        visited[idx]=true;
        s.addSegment(segs[idx]);
        for(int i=0;i<segs.length;i++){
            if(!visited[i]){
                Segment a = segs[idx], b = segs[i];
                if(pointOnSegment(a,b.x1,b.y1) || pointOnSegment(a,b.x2,b.y2) ||
                   pointOnSegment(b,a.x1,a.y1) || pointOnSegment(b,a.x2,a.y2)){
                    dfs(i,segs,visited,s);
                }
            }
        }
    }

    static boolean pointOnSegment(Segment s, int x, int y){
        return x>=Math.min(s.x1,s.x2) && x<=Math.max(s.x1,s.x2) &&
               y>=Math.min(s.y1,s.y2) && y<=Math.max(s.y1,s.y2) &&
               (slope(s.x1,s.y1,s.x2,s.y2)==slope(s.x1,s.y1,x,y));
    }

    static double slope(int x1,int y1,int x2,int y2){
        if(x1==x2) return Double.POSITIVE_INFINITY;
        return (double)(y2-y1)/(x2-x1);
    }
}

File: TCS-CODEVITA/Season-13/Round-1/Stellar-Journey/solution.py
from collections import deque

# Check if point lies on a segment
def point_on_segment(pt, seg):
    (x1,y1),(x2,y2) = seg
    if x1 == x2:  # vertical
        return pt[0] == x1 and min(y1, y2) <= pt[1] <= max(y1, y2)
    elif y1 == y2:  # horizontal
        return pt[1] == y1 and min(x1, x2) <= pt[0] <= max(x1, x2)
    else:  # diagonal
        dx = x2 - x1
        dy = y2 - y1
        if dx == 0 or dy == 0:
            return False
        return (pt[0] - x1) * dy == (pt[1] - y1) * dx and min(x1,x2) <= pt[0] <= max(x1,x2) and min(y1,y2) <= pt[1] <= max(y1,y2)

# Check if point is in star
def point_in_star(pt, star):
    return any(point_on_segment(pt, seg) for seg in star)

# Check if two segments intersect (bounding-box approximation)
def intersects(seg1, seg2):
    (x1,y1),(x2,y2) = seg1
    (x3,y3),(x4,y4) = seg2
    if max(x1,x2) < min(x3,x4) or max(x3,x4) < min(x1,x2):
        return False
    if max(y1,y2) < min(y3,y4) or max(y3,y4) < min(y1,y2):
        return False
    return True

# Read input
n = int(input())
segments = [tuple(map(int, input().split())) for _ in range(n)]
segments = [((x1,y1),(x2,y2)) for x1,y1,x2,y2 in segments]
start = tuple(map(int, input().split()))
end = tuple(map(int, input().split()))

# Step 1: Cluster segments into stars
stars = []
visited = [False]*n

for i in range(n):
    if visited[i]:
        continue
    queue = [i]
    star = []
    visited[i] = True
    while queue:
        idx = queue.pop()
        star.append(segments[idx])
        for j in range(n):
            if not visited[j] and intersects(segments[idx], segments[j]):
                visited[j] = True
                queue.append(j)
    stars.append(star)

# Step 2: Find stars containing start and end points
start_star = next((i for i, s in enumerate(stars) if point_in_star(start, s)), -1)
end_star = next((i for i, s in enumerate(stars) if point_in_star(end, s)), -1)

if start_star == -1 or end_star == -1:
    print("Impossible")
    exit()

# Step 3: Build connectivity graph
graph = {i: set() for i in range(len(stars))}
for i in range(len(stars)):
    for j in range(i+1, len(stars)):
        if any(intersects(seg1, seg2) for seg1 in stars[i] for seg2 in stars[j]):
            graph[i].add(j)
            graph[j].add(i)

# Step 4: BFS for minimum stars
queue = deque([(start_star, 1)])
visited_stars = set()

while queue:
    current, count = queue.popleft()
    if current == end_star:
        print(count)
        exit()
    visited_stars.add(current)
    for neighbor in graph.get(current, []):
        if neighbor not in visited_stars:
            queue.append((neighbor, count+1))

print("Impossible")
